### PR TITLE
bump `monty` to use the `monty.json` import speedup patch, add import time regression test, lazy load some rarely used but costly modules

### LINKED ISSUE #3793: import `monty.json` slowing down core import
### Summary

- `from monty.json import xxx` significantly slowly down core modules, as `torch` is eagerly imported
- [ ]  Might need to look into other commonly used modules

For example:
```bash
python -X importtime -c "from pymatgen.core.structure import Structure" 2> pmg.log && tuna pmg.log
```

With `torch` installed (required by optional `matgl`):
image

After removing `torch`:

image

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- Organize optional dependencies in `pyproject.toml` to avoid duplicate
- Bump `sympy >= 1.3` to resolve https://github.com/materialsproject/pymatgen/pull/4116#issuecomment-2428308320
- Avoid `from numpy.testing import assert_allclose` in production code

---

- bump `monty` to use [the latest json import speedup patch](https://github.com/materialsvirtuallab/monty/pull/713), partially fix #3793
- Have a quick look at other important modules (on cover those import other 3rd-party/non-core-pmg modules in this PR) and add import time test (profile: `python -X importtime -c "from pymatgen.core.structure import Structure" 2> pmg.log && tuna pmg.log`):
    - [x] core.bonds: image
    - [x] core.composition: image
    - [x] core.interface (**need attention**)
    - [x] core.ion: image
    - [x] core.lattice: image
    - [x] core.operations: image
    - [x] core.periodic_table: image
    - [x] core.sites: image
    - [x] core.spectrum: image
    - [x] core.structure
    - [x] core.surface: image
    - [x] core.tensors: image
    - [x] core.trajectory (`AseAtomsAdaptor` lazy imported)
    - [x] io.vasp.inputs image
    - [x] io.vasp.outputs: image

- [For a follow up PR] looks like `scipy` across `core` need special attention

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,40 +87,34 @@ Issues = "https://github.com/materialsproject/pymatgen/issues"
 Pypi = "https://pypi.org/project/pymatgen"
 
 [project.optional-dependencies]
+abinit = ["netcdf4>=1.6.5"]
 ase = ["ase>=3.23.0"]
-# tblite only support Python 3.12+ through conda-forge
-# https://github.com/tblite/tblite/issues/175
-tblite = ["tblite[ase]>=0.3.0; python_version<'3.12'"]
-vis = ["vtk>=6.0.0"]
-abinit = ["netcdf4>=1.7.1"]
```
Comment (DanielYang59) [inline pyproject.toml:95]: this separate `abinit` extra and the `optional` extra requires different versions of `netcdf4`: https://github.com/materialsproject/pymatgen/blob/aa37e99fa67a0e9f996b645be26c12a8f42a4a98/pyproject.toml#L113

pin `netcdf4>=1.7.1` might lead to CI failure in Ubuntu (which I cannot recreate locally)

**Comment by mkhorton** [comment]
> @mkhorton Can I lazy import `sympy` (it's used by only one method `symmetry.settings.JonesFaithfulTransformation.parse_transformation_string` and would give us a ~15% speed up on import `core.Structure`)?

Please! This method is very rarely used, lazy import preferred.

**Comment by mkhorton** [comment]
Thank you for the profiling, let me know when ready for review.

File: src/pymatgen/symmetry/settings.py [hunk 2]
```diff
@@ -100,6 +98,10 @@ def parse_transformation_string(
         Returns:
             tuple[list[list[float]] | np.ndarray, list[float]]: transformation matrix & vector
         """
+        # Import sympy is expensive
```
Comment (DanielYang59) [inline src/pymatgen/symmetry/settings.py:101]: ## import profile of `core.structure`

### Current master branch with `torch`

image


### After `monty.json` patch

image


### After lazily load `sympy`

`sympy` is required by `symmetry.settings.JonesFaithfulTransformation.parse_transformation_string`, and reduce import time by about 15%, sounds like a good deal

image


### After lazily load `scipy.spatial.Voronoi` (reverted)

Usage of `scipy` seem very common across `pymatgen` too, doesn't bring much benefit, reverted.

File: src/pymatgen/core/interface.py [hunk 3]
```diff
@@ -2784,10 +2783,16 @@ def from_slabs(
             substrate_slab = substrate_slab.get_orthogonal_c_slab()
         if isinstance(film_slab, Slab):
             film_slab = film_slab.get_orthogonal_c_slab()
-        assert_allclose(film_slab.lattice.alpha, 90, 0.1)
```
Comment (DanielYang59) [inline src/pymatgen/core/interface.py:2787]: `math.isclose` is much faster (~ 500x) for comparing scalar than `assert_allclose`.


```python
import math
from numpy.testing import assert_allclose
import timeit


def run_tests():
    a = 0.123456789
    b = 0.123456789

    rtol = 1e-5
    atol = 1e-8

    assert math.isclose(a, b, rel_tol=rtol, abs_tol=atol)
    assert_allclose(a, b, rtol=rtol, atol=atol)

    # Measure performance in milliseconds
    num_runs = 1000000
    isclose_time = timeit.timeit(lambda: math.isclose(a, b, rel_tol=rtol, abs_tol=atol), number=num_runs) * 1000
    assert_allclose_time = timeit.timeit(lambda: assert_allclose(a, b, rtol=rtol, atol=atol), number=num_runs) * 1000

    print(f"\nPerformance results over {num_runs} runs with single float values:")
    print(f"math.isclose time: {isclose_time:.3f} ms")
    print(f"assert_allclose time: {assert_allclose_time:.3f} ms")
    print(f"Speedup: {assert_allclose_time / isclose_time:.2f}x")

run_tests()
```

Gives:
```
Performance results over 1000000 runs with single float values:
math.isclose time: 37.630 ms
assert_allclose time: 19495.991 ms
Speedup: 518.09x
```

File: src/pymatgen/analysis/interfaces/coherent_interfaces.py [hunk 4]
```diff
@@ -101,21 +101,15 @@ def _find_matches(self) -> None:
         for match in self.zsl_matches:
             xform = get_2d_transform(film_vectors, match.film_vectors)
             strain, _rot = polar(xform)
-            assert_allclose(
```
Comment (DanielYang59) [inline src/pymatgen/analysis/interfaces/coherent_interfaces.py:104]: Similarly, `np.testing.assert_allclose` is not suitable for production code (~ 2.5x slower than `np.isclose`), though it provides more detailed debug info for testing.

Note the default tolerances are a bit stricter for `assert_allclose`, this commit also migrated the tolerances as is without changing in behaviour.
> testing.assert_allclose(actual, desired, rtol=1e-07, atol=0, equal_nan=True, err_msg='', verbose=True, *, strict=False)

> numpy.allclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False)

Test script:
```python
import numpy as np
from numpy.testing import assert_allclose
import timeit


def run_tests():
    # Array sizes to test
    sizes = [(10, 10), (100, 100), (1000, 1000), (10000, 10000)]

    rtol = 1e-5
    atol = 1e-8

    # Run tests for each size
    for size in sizes:
        arr1 = np.random.rand(*size)
        arr2 = np.copy(arr1)  # Exact copy of arr1

        assert_allclose(arr1, arr2, rtol=rtol, atol=atol)
        assert np.allclose(arr1, arr2, rtol=rtol, atol=atol)

        # Measure performance in milliseconds
        num_runs = 10
        assert_time = timeit.timeit(lambda: assert_allclose(arr1, arr2, rtol=rtol, atol=atol), number=num_runs) * 1000
        allclose_time = timeit.timeit(lambda: np.allclose(arr1, arr2, rtol=rtol, atol=atol), number=num_runs) * 1000

        print(f"\nPerformance results for array size {size} over {num_runs} runs:")
        print(f"assert_allclose time: {assert_time:.3f} ms")
        print(f"np.allclose time: {allclose_time:.3f} ms")
        print(f"Speedup: {assert_time / allclose_time:.2f}x")

run_tests()
```

I got:

```
Performance results for array size (10, 10) over 10 runs:
assert_allclose time: 0.308 ms
np.allclose time: 0.102 ms
Speedup: 3.03x

Performance results for array size (100, 100) over 10 runs:
assert_allclose time: 0.563 ms
np.allclose time: 0.244 ms
Speedup: 2.30x

Performance results for array size (1000, 1000) over 10 runs:
assert_allclose time: 55.887 ms
np.allclose time: 22.179 ms
Speedup: 2.52x

Performance results for array size (10000, 10000) over 10 runs:
assert_allclose time: 12374.578 ms
np.allclose time: 2916.093 ms
Speedup: 4.24x
```

File: src/pymatgen/core/trajectory.py [hunk 5]
```diff
@@ -580,9 +579,12 @@ def from_file(cls, filename: str | Path, constant_lattice: bool = True, **kwargs
             try:
                 from ase.io.trajectory import Trajectory as AseTrajectory
 
+                from pymatgen.io.ase import AseAtomsAdaptor
```
Comment (DanielYang59) [inline src/pymatgen/core/trajectory.py:582]: `AseAtomsAdaptor` is only used in one of the many try-except branches (other branches also lazy import the corresponding module)

https://github.com/materialsproject/pymatgen/blob/6992aeef9847b03da81c80dabe27c84010e28abc/src/pymatgen/core/trajectory.py#L569-L593

File: src/pymatgen/analysis/adsorption.py [hunk 6]
```diff
@@ -664,6 +662,10 @@ def plot_slab(
         decay (float): how the alpha-value decays along the z-axis
         inverse (bool): invert z axis to plot opposite surface
     """
+    # Expensive import (PR4128)
+    from matplotlib import patches
```
Comment (DanielYang59) [inline src/pymatgen/analysis/adsorption.py:666]: `matplotlib` is almost certainly not a core module for pymatgen.core, and it incurred significant import overhead to `core.interface`:
image

After: 
image

**Comment by DanielYang59** [comment]
@mkhorton I believe this PR is already for view, let me know if you have any comment, thank you!

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -82,7 +82,7 @@ jobs:
           # TODO1 (use uv over pip) uv install torch is flaky, track #3826
           # TODO2 (pin torch version): DGL library (matgl) doesn't support torch > 2.2.1,
           # see: https://discuss.dgl.ai/t/filenotfounderror-cannot-find-dgl-c-graphbolt-library/4302
-          pip install torch==2.2.1
+          pip install --timeout=300 torch==2.2.1
```
Comment (DanielYang59) [inline .github/workflows/test.yml:85]: it seems even `pip install torch` is quite flaky as well, and I guess [it's not to be fixed in the short term](https://github.com/pypi/warehouse/issues/3826#issuecomment-384588312):

> Occasional timeouts will happen.

They recommended bumping the timeout instead (default is 15 s)

**Comment by janosh** [comment]
thanks a lot for your extensive profiling work @DanielYang59! i expected there would be a some low-hanging fruit but not nearly as much as you found. you went above and beyond here. 👍 
will be big help in resolving https://github.com/janosh/pymatviz/issues/209

**Comment by DanielYang59** [comment]
thanks @shyuep this hopefully would significantly speedup import

**Pull request merged**